core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -533,10 +542,14 @@ boolean deleteValueFinish(ThreadContext ctx) {
         }
 
         // marking the delete bit in the key's off-heap header (only one true setter gets result TRUE)
-        // The marking happens only when no lock is taken, otherwise busy waits
+        // (Marking happens prior to reference being marked is deleted,
+        // thus if reference is deleted the off-heap is already marked.)
+        // The marking the delete bit happens only when no lock is taken, otherwise busy waits
         // if the version is already different result is RETRY, if already deleted - FALSE
         // we continue anyway, therefore disregard the logicalDelete() output
-        ctx.key.s.logicalDelete();
+        if (!keysMemoryManager.isReferenceDeleted(expectedKeyReference)) {
```

Comment:
We don't need to check `isReferenceDeleted()` twice in the same reference. It doesn't change.
What is the purpose of making this check here and not later (right before the CAS below)? Is this that we logically delete the key before we delete the value?
If so, please state that in the comment.